conformance/tests/httproute-cors.go

```diff
@@ -0,0 +1,118 @@
+package tests
+
+import (
+	"testing"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/gateway-api/conformance/utils/http"
+	"sigs.k8s.io/gateway-api/conformance/utils/kubernetes"
+	"sigs.k8s.io/gateway-api/conformance/utils/suite"
+	"sigs.k8s.io/gateway-api/pkg/features"
+)
+
+func init() {
+	ConformanceTests = append(ConformanceTests, HTTPRouteCORS)
+}
+
+var HTTPRouteCORS = suite.ConformanceTest{
+	ShortName:   "HTTPRouteCORS",
+	Description: "An HTTPRoute with cors filter",
+	Features: []features.FeatureName{
+		features.SupportGateway,
+		features.SupportHTTPRoute,
+		features.SupportHTTPRouteCORS,
+	},
+	Manifests: []string{"tests/httproute-cors.yaml"},
+	Test: func(t *testing.T, suite *suite.ConformanceTestSuite) {
+		ns := "gateway-conformance-infra"
+		routeNN := types.NamespacedName{Name: "cors", Namespace: ns}
+		gwNN := types.NamespacedName{Name: "same-namespace", Namespace: ns}
+		gwAddr := kubernetes.GatewayAndHTTPRoutesMustBeAccepted(t, suite.Client, suite.TimeoutConfig, suite.ControllerName, kubernetes.NewGatewayRef(gwNN), routeNN)
+		kubernetes.HTTPRouteMustHaveResolvedRefsConditionsTrue(t, suite.Client, suite.TimeoutConfig, routeNN, gwNN)
+
+		testCases := []http.ExpectedResponse{{
+			TestCaseName: "GET request",
+			Request: http.Request{
+				Path: "/one",
+				Headers: map[string]string{
+					"Origin": "http://example.com",
+				},
+			},
+			Response: http.Response{
+				StatusCode: 200,
+				Headers: map[string]string{
+					"Access-Control-Allow-Origin": "http://example.com",
+				},
+			},
+			Backend:   "infra-backend-v1",
+			Namespace: ns,
+		}, {
+			TestCaseName: "CORS preflight request with request method and headers",
+			Request: http.Request{
+				Method: "OPTIONS",
+				Path:   "/one",
+				Headers: map[string]string{
+					"Origin":                         "http://one.example.com",
+					"Access-Control-Request-Method":  "POST",
+					"Access-Control-Request-Headers": "Accept,Content-Type",
+				},
+			},
+			ExpectedRequest: &http.ExpectedRequest{
+				Request: http.Request{
+					Method: "OPTIONS",
+				},
+			},
+			Response: http.Response{
+				StatusCode: 200,
+				Headers: map[string]string{
+					"Access-Control-Allow-Origin":  "http://one.example.com",
+					"Access-Control-Allow-Methods": "GET,HEAD,POST",
+					"Access-Control-Allow-Headers": "Accept,Accept-Language,Content-Language,Content-Type,Range",
+					"Access-Control-Max-Age":       "5",
+				},
+				AbsentHeaders: []string{
+					"Access-Control-Allow-Credentials",
+					"Access-Control-Expose-Headers",
+				},
+			},
+		}, {
+			TestCaseName: "CORS preflight request with request method",
+			Request: http.Request{
+				Method: "OPTIONS",
+				Path:   "/two",
+				Headers: map[string]string{
+					"Origin":                        "http://two.example.com",
+					"Access-Control-Request-Method": "DELETE",
+				},
+			},
+			ExpectedRequest: &http.ExpectedRequest{
```

Comment:
Isn't ExpectedRequest the request that hits the backend? Wouldn't this not hit the backend at all?

Reply:
Yes, exactly. However `ExpectedRequest: nil` implicitly sets the actual request as expected request, which isn't correct for this case.

Also. when I tried to set the expected request to be empty with `ExpectedRequest: &http.ExpectedRequest{}`, it still didn't work. This is because the method of the captured request is set to the method of the actual request, even if there there is no captured request at all (https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/utils/roundtripper/roundtripper.go#L252).